h5py/_hl/files.py

```diff
@@ -254,8 +254,14 @@ def driver(self):
     @with_phil
     def mode(self):
         """ Python mode used to open file """
-        return {h5f.ACC_RDONLY: 'r',
-                h5f.ACC_RDWR: 'r+'}.get(self.id.get_intent())
+        if swmr_support:
+            intent = self.id.get_intent() & (h5f.ACC_SWMR_WRITE | h5f.ACC_RDWR)
+        else:
+            intent = self.id.get_intent() & h5f.ACC_RDWR
+        if 0 < intent:
+            return 'r+'
+        else:
+            return 'r'
```

Comment:
I think we can condense this a bit:

```suggestion
        write_intent = h5f.ACC_RDWR
        if swmr_support:
            write_intent |= h5f.ACC_SWMR_WRITE
        return 'r+' if self.id.get_intent() & write_intent else 'r'
```